Use docker compose down command for container removal. And minor impr…

tobwiens opened the pull request:
…ovements.

Increase the script-engine version to 0.2.0.

The docker compose stop and docker compose remove command were both too slow. So the ProActive node killed the thread during the removal process.

The docker compose down command is now utilised to remove and kill the container within one go, which increases the likelihood of the container actually being removed.

The DockerComposeScriptEngineFactory got its own ProcessBuilder service class.

The DockerComposeScriptEngineFactory got a standard constructor with which it can be initialized at during the node startup.

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/bindings/MapBindingsAdder.java:
```diff
@@ -34,23 +34,21 @@
  */
 package jsr223.docker.compose.bindings;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Rename this package name to match the regular expression '^[a-z]+(.[a-z][a-z0-9]_)_$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00120)

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/DockerComposeScriptEngineFactory.java:
```diff
@@ -1,41 +1,42 @@
 package jsr223.docker.compose;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Rename this package name to match the regular expression '^[a-z]+(.[a-z][a-z0-9]_)_$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00120)

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/DockerComposeScriptEngineFactory.java:
```diff
@@ -1,41 +1,42 @@
-    private DockerComposeVersionGetter dockerComposeVersionGetter;
-
     // Script engine parameters
     private static final String NAME = "docker-compose";
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal "docker-compose" 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1192)

sonarqube-activeeon commented on src/main/java/processbuilder/utils/ProcessBuilderUtilities.java:
```diff
@@ -18,7 +18,7 @@
-    public void attachToInputStream(final Reader source, final Writer attachedSink) {
+    private void attachToInputStream(final Reader source, final Writer attachedSink) {
         new Thread() {
             public void run() {
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add the "@Override" annotation above this method signature [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1161)

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/DockerComposeScriptEngine.java:
```diff
@@ -1,39 +1,33 @@
 package jsr223.docker.compose;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Rename this package name to match the regular expression '^[a-z]+(.[a-z][a-z0-9]_)_$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00120)

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/DockerComposeScriptEngine.java:
```diff
@@ -96,41 +90,37 @@ public Object eval(String script, ScriptContext context) throws ScriptException
+                // Reset thread's interrupt flag
+                Thread.interrupted();
+                stopAndRemoveContainers().waitFor();
+            } catch (Exception e) {
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/DockerComposeScriptEngine.java:
```diff
@@ -96,41 +90,37 @@ public Object eval(String script, ScriptContext context) throws ScriptException
+                Thread.interrupted();
+                stopAndRemoveContainers().waitFor();
+            } catch (Exception e) {
+                e.printStackTrace();
```
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Use a logger to log this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1148)

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/DockerComposeScriptEngine.java:
```diff
@@ -96,41 +90,37 @@ public Object eval(String script, ScriptContext context) throws ScriptException
+    }
+
 
     // TODO: Test
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1135)

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/utils/DockerComposePropertyLoader.java:
```diff
@@ -46,20 +46,20 @@
+    private final Properties properties;
 
-    protected String configurationFile = "config/scriptengines/docker-compose.properties";
+    private final String configurationFile = "config/scriptengines/docker-compose.properties";
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Make this final field static too. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1170)

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/DockerComposeCommandCreator.java:
```diff
@@ -1,47 +1,41 @@
 package jsr223.docker.compose;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Rename this package name to match the regular expression '^[a-z]+(.[a-z][a-z0-9]_)_$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00120)

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/DockerComposeCommandCreator.java:
```diff
@@ -1,47 +1,41 @@
+    @Getter
+    private final String yamlFileName = "docker-compose.yml";
+    @Getter
+    private final String filenameArgument = "-f";
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Make this final field static too. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1170)

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/DockerComposeCommandCreator.java:
```diff
@@ -1,47 +1,41 @@
+    @Getter
+    private final String filenameArgument = "-f";
+    @Getter
+    private final String setupContainerArgument = "up";
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Make this final field static too. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1170)

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/DockerComposeCommandCreator.java:
```diff
@@ -1,47 +1,41 @@
+    @Getter
+    private final String setupContainerArgument = "up";
+    @Getter
+    private final String stopAndRemoveContainerArgument = "down";
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Make this final field static too. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1170)

sonarqube-activeeon commented:
SonarQube analysis reported 24 issues
* ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) 2 critical
* ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) 8 major
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 13 minor
* ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) 1 info

Watch the comments in this conversation to review them.

#### 9 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) [DockerComposeScriptEngine.java#L47](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/5649516484734a2c040ab171b8f1549f9a711357/src/main/java/jsr223/docker/compose/DockerComposeScriptEngine.java#L47): Use a logger to log this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1148)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [DockerComposeScriptEngine.java#L40](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/5649516484734a2c040ab171b8f1549f9a711357/src/main/java/jsr223/docker/compose/DockerComposeScriptEngine.java#L40): Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [DockerComposeScriptEngine.java#L41](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/5649516484734a2c040ab171b8f1549f9a711357/src/main/java/jsr223/docker/compose/DockerComposeScriptEngine.java#L41): Replace this usage of System.out or System.err by a logger. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS106)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [DockerComposeScriptEngine.java#L43](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/5649516484734a2c040ab171b8f1549f9a711357/src/main/java/jsr223/docker/compose/DockerComposeScriptEngine.java#L43): Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [DockerComposeScriptEngine.java#L44](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/5649516484734a2c040ab171b8f1549f9a711357/src/main/java/jsr223/docker/compose/DockerComposeScriptEngine.java#L44): Replace this usage of System.out or System.err by a logger. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS106)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [DockerComposePropertyLoader.java#L95](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/5649516484734a2c040ab171b8f1549f9a711357/src/main/java/jsr223/docker/compose/utils/DockerComposePropertyLoader.java#L95): Add a private constructor to hide the implicit public one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1118)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [ProcessBuilderUtilities.java#L26](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/5649516484734a2c040ab171b8f1549f9a711357/src/main/java/processbuilder/utils/ProcessBuilderUtilities.java#L26): Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)
1. ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [DockerComposePropertyLoader.java#L35](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/5649516484734a2c040ab171b8f1549f9a711357/src/main/java/jsr223/docker/compose/utils/DockerComposePropertyLoader.java#L35): Rename this package name to match the regular expression '^[a-z]+(\.[a-z][a-z0-9]*)*$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00120)
1. ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [DockerComposeVersionGetter.java#L1](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/5649516484734a2c040ab171b8f1549f9a711357/src/main/java/jsr223/docker/compose/utils/DockerComposeVersionGetter.java#L1): Rename this package name to match the regular expression '^[a-z]+(\.[a-z][a-z0-9]*)*$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00120)

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/DockerComposeCommandCreator.java:
```diff
@@ -1,47 +1,41 @@
+    @Getter
+    private final String stopAndRemoveContainerArgument = "down";
+    @Getter
+    private final String volumesArgument = "--volumes";
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Make this final field static too. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1170)

yinan-liu replied:
lombok @Data annotation on class level can be used instead @Getter on each field. if the field is final, only getter methods will be generated by lombok. This can save you some code.

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/DockerComposeCommandCreator.java:
```diff
@@ -1,47 +1,41 @@
-        List<String> command = new ArrayList<>();
-        addSudoAndDockerComposeCommand(command);
+    @Getter
+    private final String yamlFileName = "docker-compose.yml";
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Make this final field static too. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1170)

sonarqube-activeeon commented:
SonarQube analysis reported 23 issues
* ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) 1 critical
* ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) 8 major
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 13 minor
* ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) 1 info

Watch the comments in this conversation to review them.

#### 9 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) [DockerComposeScriptEngine.java#L47](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/eb6d864f497014aba5b09f6cc6e9401017d3e665/src/main/java/jsr223/docker/compose/DockerComposeScriptEngine.java#L47): Use a logger to log this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1148)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [DockerComposeScriptEngine.java#L40](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/eb6d864f497014aba5b09f6cc6e9401017d3e665/src/main/java/jsr223/docker/compose/DockerComposeScriptEngine.java#L40): Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [DockerComposeScriptEngine.java#L41](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/eb6d864f497014aba5b09f6cc6e9401017d3e665/src/main/java/jsr223/docker/compose/DockerComposeScriptEngine.java#L41): Replace this usage of System.out or System.err by a logger. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS106)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [DockerComposeScriptEngine.java#L43](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/eb6d864f497014aba5b09f6cc6e9401017d3e665/src/main/java/jsr223/docker/compose/DockerComposeScriptEngine.java#L43): Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [DockerComposeScriptEngine.java#L44](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/eb6d864f497014aba5b09f6cc6e9401017d3e665/src/main/java/jsr223/docker/compose/DockerComposeScriptEngine.java#L44): Replace this usage of System.out or System.err by a logger. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS106)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [DockerComposePropertyLoader.java#L95](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/eb6d864f497014aba5b09f6cc6e9401017d3e665/src/main/java/jsr223/docker/compose/utils/DockerComposePropertyLoader.java#L95): Add a private constructor to hide the implicit public one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1118)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [ProcessBuilderUtilities.java#L26](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/eb6d864f497014aba5b09f6cc6e9401017d3e665/src/main/java/processbuilder/utils/ProcessBuilderUtilities.java#L26): Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)
1. ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [DockerComposePropertyLoader.java#L35](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/eb6d864f497014aba5b09f6cc6e9401017d3e665/src/main/java/jsr223/docker/compose/utils/DockerComposePropertyLoader.java#L35): Rename this package name to match the regular expression '^[a-z]+(\.[a-z][a-z0-9]*)*$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00120)
1. ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [DockerComposeVersionGetter.java#L1](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/eb6d864f497014aba5b09f6cc6e9401017d3e665/src/main/java/jsr223/docker/compose/utils/DockerComposeVersionGetter.java#L1): Rename this package name to match the regular expression '^[a-z]+(\.[a-z][a-z0-9]*)*$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00120)

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/utils/DockerComposePropertyLoader.java:
```diff
@@ -46,29 +46,29 @@
+    private final Properties properties;
 
-    protected String configurationFile = "config/scriptengines/docker-compose.properties";
+    private final static String CONFIGURATION_FILE = "config/scriptengines/docker-compose.properties";
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Reorder the modifiers to comply with the Java Language Specification. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AModifiersOrderCheck)

sonarqube-activeeon commented:
SonarQube analysis reported 18 issues
* ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) 1 critical
* ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) 8 major
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 8 minor
* ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) 1 info

Watch the comments in this conversation to review them.

#### 9 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) [DockerComposeScriptEngine.java#L47](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/dab0ec4dbab24c60171d62e7bb452c07e7e7b422/src/main/java/jsr223/docker/compose/DockerComposeScriptEngine.java#L47): Use a logger to log this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1148)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [DockerComposeScriptEngine.java#L40](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/dab0ec4dbab24c60171d62e7bb452c07e7e7b422/src/main/java/jsr223/docker/compose/DockerComposeScriptEngine.java#L40): Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [DockerComposeScriptEngine.java#L41](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/dab0ec4dbab24c60171d62e7bb452c07e7e7b422/src/main/java/jsr223/docker/compose/DockerComposeScriptEngine.java#L41): Replace this usage of System.out or System.err by a logger. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS106)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [DockerComposeScriptEngine.java#L43](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/dab0ec4dbab24c60171d62e7bb452c07e7e7b422/src/main/java/jsr223/docker/compose/DockerComposeScriptEngine.java#L43): Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [DockerComposeScriptEngine.java#L44](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/dab0ec4dbab24c60171d62e7bb452c07e7e7b422/src/main/java/jsr223/docker/compose/DockerComposeScriptEngine.java#L44): Replace this usage of System.out or System.err by a logger. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS106)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [DockerComposePropertyLoader.java#L95](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/dab0ec4dbab24c60171d62e7bb452c07e7e7b422/src/main/java/jsr223/docker/compose/utils/DockerComposePropertyLoader.java#L95): Add a private constructor to hide the implicit public one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1118)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [ProcessBuilderUtilities.java#L26](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/dab0ec4dbab24c60171d62e7bb452c07e7e7b422/src/main/java/processbuilder/utils/ProcessBuilderUtilities.java#L26): Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)
1. ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [DockerComposePropertyLoader.java#L35](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/dab0ec4dbab24c60171d62e7bb452c07e7e7b422/src/main/java/jsr223/docker/compose/utils/DockerComposePropertyLoader.java#L35): Rename this package name to match the regular expression '^[a-z]+(\.[a-z][a-z0-9]*)*$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00120)
1. ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [DockerComposeVersionGetter.java#L1](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/dab0ec4dbab24c60171d62e7bb452c07e7e7b422/src/main/java/jsr223/docker/compose/utils/DockerComposeVersionGetter.java#L1): Rename this package name to match the regular expression '^[a-z]+(\.[a-z][a-z0-9]*)*$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00120)

yinan-liu commented on src/main/java/jsr223/docker/compose/utils/DockerComposePropertyLoader.java:
```diff
@@ -45,30 +46,25 @@
+    private final String dockerComposeCommand;
+    @Getter
+    private final String sudoCommand;
+    @Getter
```
same, use @Data should be fine

tobwiens replied:
I don't want to have private final Properties properties accessible from outside, because one can alter the Properties instance. That would make the class mutable.

yinan-liu replied:
i didn't get you, then why you set @Getter on these fields ?  use @Data at class level, when the attributes are final, lombok will generate only getter method, not setter.

tobwiens replied:
I don't want getter methods for all members. As an example:  a HashMap, if anyone is able to get a HashMap reference, then one can change the content of that HashMap from the outside.

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/utils/DockerComposePropertyLoader.java:
```diff
@@ -45,30 +46,25 @@
-    protected String configurationFile = "config/scriptengines/docker-compose.properties";
-
-    private DockerComposePropertyLoader()  {
+    private final static String CONFIGURATION_FILE = "config/scriptengines/docker-compose.properties";
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Reorder the modifiers to comply with the Java Language Specification. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AModifiersOrderCheck)

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/utils/DockerComposePropertyLoader.java:
```diff
@@ -35,6 +35,7 @@
 package jsr223.docker.compose.utils;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Rename this package name to match the regular expression '^[a-z]+(.[a-z][a-z0-9]_)_$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00120)

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/utils/DockerComposePropertyLoader.java:
```diff
@@ -83,20 +79,18 @@ private DockerComposePropertyLoader()  {
+     * or the first access to DockerComposePropertyLoaderHolder.INSTANCE, not before.
+     **/
+    @NoArgsConstructor(access = AccessLevel.PRIVATE)
     private static class DockerComposePropertyLoaderHolder {
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add a private constructor to hide the implicit public one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1118)

sonarqube-activeeon commented:
SonarQube analysis reported 13 issues
* ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) 4 major
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 8 minor
* ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) 1 info

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [ProcessBuilderUtilities.java#L26](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/0577f19f65274c685b144af1d600b9e8a9914f4e/src/main/java/processbuilder/utils/ProcessBuilderUtilities.java#L26): Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)
1. ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [DockerComposeVersionGetter.java#L1](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/0577f19f65274c685b144af1d600b9e8a9914f4e/src/main/java/jsr223/docker/compose/utils/DockerComposeVersionGetter.java#L1): Rename this package name to match the regular expression '^[a-z]+(\.[a-z][a-z0-9]*)*$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00120)

sonarqube-activeeon commented:
SonarQube analysis reported 12 issues
* ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) 3 major
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 8 minor
* ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [DockerComposeVersionGetter.java#L1](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/250003316d69777dfd7bc04402f06a68c615c817/src/main/java/jsr223/docker/compose/utils/DockerComposeVersionGetter.java#L1): Rename this package name to match the regular expression '^[a-z]+(\.[a-z][a-z0-9]*)*$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00120)

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/utils/DockerComposePropertyLoader.java:
```diff
@@ -35,6 +35,7 @@
 package jsr223.docker.compose.utils;
 
 import lombok.AccessLevel;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'lombok.AccessLevel'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/utils/DockerComposePropertyLoader.java:
```diff
@@ -35,6 +35,7 @@
 
 import lombok.AccessLevel;
+import lombok.Getter;
 import lombok.NoArgsConstructor;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'lombok.NoArgsConstructor'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

sonarqube-activeeon commented:
SonarQube analysis reported 13 issues
* ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) 2 major
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 10 minor
* ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [DockerComposeVersionGetter.java#L1](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/999318130c030fd0bbc44494b018973cf7f86c3c/src/main/java/jsr223/docker/compose/utils/DockerComposeVersionGetter.java#L1): Rename this package name to match the regular expression '^[a-z]+(\.[a-z][a-z0-9]*)*$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00120)

sonarqube-activeeon commented:
SonarQube analysis reported 12 issues
* ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) 2 major
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 9 minor
* ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [DockerComposeVersionGetter.java#L1](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/94c9741937e401ffd466a4292841b6fb54501179/src/main/java/jsr223/docker/compose/utils/DockerComposeVersionGetter.java#L1): Rename this package name to match the regular expression '^[a-z]+(\.[a-z][a-z0-9]*)*$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00120)

tobwiens committed:
Use docker compose down command for container removal.

And minor improvements.

Increase the script-engine version to 0.2.0.

The docker compose stop and docker compose remove command were both too slow. So the ProActive node killed the thread during the removal process.

The docker compose down command is now utilised to remove and kill the container within one go, which increases the likelihood of the container actually being removed.

The DockerComposeScriptEngineFactory got its own ProcessBuilder service class.

The DockerComposeScriptEngineFactory got a standard constructor with which it can be initialized at during the node startup.

sonarqube-activeeon commented:
SonarQube analysis reported 12 issues
* ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) 2 major
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 9 minor
* ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [DockerComposeVersionGetter.java#L1](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/ab3ba9eeb9a06c5bc66396d81bab3c5897e2c107/src/main/java/jsr223/docker/compose/utils/DockerComposeVersionGetter.java#L1): Rename this package name to match the regular expression '^[a-z]+(\.[a-z][a-z0-9]*)*$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00120)

yinan-liu commented on src/main/java/jsr223/docker/compose/DockerComposeScriptEngine.java:
```diff
@@ -43,14 +37,14 @@ public DockerComposeScriptEngine() {
-                    ". Docker Compose script engine is functional but logging is disabled." +
-                    "Stacktrace is: ");
-            e.printStackTrace();
+        } catch (NullPointerException e) { //NOSONAR
```
this part needs to be refactored, what do you think?

tobwiens replied:
Refactored :+1:

tobwiens committed:
Refactor log4j initialization.

Add service class to initialise log4j configuration.

sonarqube-activeeon commented on src/main/java/jsr223/docker/compose/utils/Log4jConfigurationLoader.java:
```diff
@@ -0,0 +1,66 @@
+ *
+ *  * $$ACTIVEEON_CONTRIBUTOR$$
+ */
+package jsr223.docker.compose.utils;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Rename this package name to match the regular expression '^[a-z]+(.[a-z][a-z0-9]_)_$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00120)

sonarqube-activeeon commented:
SonarQube analysis reported 13 issues
* ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) 2 major
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 10 minor
* ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [DockerComposeVersionGetter.java#L1](https://github.com/ow2-proactive/scriptengine-docker-compose/blob/df3acd86556644316b1943c68a96ae42fc3cdbbc/src/main/java/jsr223/docker/compose/utils/DockerComposeVersionGetter.java#L1): Rename this package name to match the regular expression '^[a-z]+(\.[a-z][a-z0-9]*)*$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00120)

tobwiens commented:
@yinan-liu Thank you for your review :). What do you think? Do you have some more comments?
